File: src/access-control/documentAccessPolicyFactory.ts
import { RawDocument } from "../document/document";
import { ResourceAccessPolicy } from "./resourceAccessPolicy";

/**
 * Interface for specifying the access policies for a document during ingestion.
 */
export interface DocumentAccessPolicyFactory {
  /**
   * Specify the access policies for a document during ingestion.
   * @param rawDocument - The RawDocument to specify policies for.
   * @returns list of ResourceAccessPolicies applicable to the document.
   */
  getAccessPolicies: (
    rawDocument: RawDocument,
  ) => Promise<ResourceAccessPolicy[]>;
}

File: python/src/semantic_retrieval/access_control/document_access_policy_factory.py
from abc import ABC, abstractmethod
from typing import List

from semantic_retrieval.access_control.resource_access_policy import (
    ResourceAccessPolicy,
)
from semantic_retrieval.document.document import RawDocument


class DocumentAccessPolicyFactory(ABC):
    @abstractmethod
    async def get_access_policies(
        self, raw_document: RawDocument
    ) -> List[ResourceAccessPolicy]:
        pass
